packages/world-modules/src/modules/erc20-puppet/ERC20Module.sol

```diff
@@ -0,0 +1,93 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.21;
+
+import { ResourceIds } from "@latticexyz/store/src/codegen/tables/ResourceIds.sol";
+import { ResourceId } from "@latticexyz/store/src/ResourceId.sol";
+import { Module } from "@latticexyz/world/src/Module.sol";
+import { WorldResourceIdLib } from "@latticexyz/world/src/WorldResourceId.sol";
+import { IBaseWorld } from "@latticexyz/world/src/codegen/interfaces/IBaseWorld.sol";
+import { InstalledModules } from "@latticexyz/world/src/codegen/tables/InstalledModules.sol";
+
+import { Puppet } from "../puppet/Puppet.sol";
+import { createPuppet } from "../puppet/createPuppet.sol";
+import { MODULE_NAME as PUPPET_MODULE_NAME } from "../puppet/constants.sol";
+
+import { MODULE_NAME, MODULE_NAMESPACE, MODULE_NAMESPACE_ID, ERC20_REGISTRY_TABLE_ID } from "./constants.sol";
+import { _allowancesTableId, _balancesTableId, _metadataTableId, _erc20SystemId } from "./utils.sol";
+import { ERC20System } from "./ERC20System.sol";
+
+import { ERC20Registry } from "./tables/ERC20Registry.sol";
+import { Balances } from "./tables/Balances.sol";
+import { Allowances } from "./tables/Allowances.sol";
+import { Metadata, MetadataData } from "./tables/Metadata.sol";
+
+contract ERC20Module is Module {
+  error ERC20Module_InvalidNamespace(bytes14 namespace);
+
+  function getName() public pure override returns (bytes16) {
+    return MODULE_NAME;
+  }
+
+  /**
+   * Register systems and tables for a new ERC20 token in a given namespace
+   */
+  function _registerERC20(bytes14 namespace) internal {
+    // Register the tables
+    Allowances.register(_allowancesTableId(namespace));
+    Balances.register(_balancesTableId(namespace));
+    Metadata.register(_metadataTableId(namespace));
+
+    // Register a new ERC20System
+    IBaseWorld(_world()).registerSystem(_erc20SystemId(namespace), new ERC20System(), true);
+  }
+
+  function _requireDependencies() internal view {
+    // If the PuppetModule is not installed yet, install it
```

Comment:
in a world where `install` becomes idempotent, can we detect+install the puppet module instead of reverting?

Reply:
Had that at some intermediate state in this PR, but it blew the module's bytecode size way over the limit (because the puppet module's bytecode had to be inlinted to the erc20 module)

Reply:
oh interesting! makes sense to go with this approach then

I wonder if we could eventually get to a place where the install just takes in the CREATE2 address of the module and we expect it to exist on chain (deployed via our deterministic deployer)

Reply:
Agree that would be nice